bookkeeper-server/src/test/java/org/apache/bookkeeper/bookie/CheckpointOnNewLedgersTest.java

```diff
@@ -0,0 +1,197 @@
+import java.io.File;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ThreadLocalRandom;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.bookkeeper.bookie.Journal.LastLogMark;
+import org.apache.bookkeeper.bookie.storage.ldb.DbLedgerStorage;
+import org.apache.bookkeeper.conf.ServerConfiguration;
+import org.apache.bookkeeper.conf.TestBKConfiguration;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+/**
+ * Test the checkpoint logic of {@link DbLedgerStorage}.
+ */
+@Slf4j
+public class CheckpointOnNewLedgersTest {
+
+    @Rule
+    public final TemporaryFolder testDir = new TemporaryFolder();
+
+    private ServerConfiguration conf;
+    private Bookie bookie;
+    private CountDownLatch getLedgerDescCalledLatch;
+    private CountDownLatch getLedgerDescWaitLatch;
+
+    @Before
+    public void setup() throws Exception {
+        File bkDir = testDir.newFolder("dbLedgerStorageCheckpointTest");
+        File curDir = Bookie.getCurrentDirectory(bkDir);
+        Bookie.checkDirectoryStructure(curDir);
+
+        int gcWaitTime = 1000;
+        conf = TestBKConfiguration.newServerConfiguration();
+        conf.setGcWaitTime(gcWaitTime);
+        conf.setLedgerStorageClass(InterleavedLedgerStorage.class.getName());
+        conf.setJournalDirsName(new String[] { bkDir.toString() });
+        conf.setLedgerDirNames(new String[] { bkDir.toString() });
+        conf.setEntryLogSizeLimit(10 * 1024);
+
+        bookie = spy(new Bookie(conf));
+        bookie.start();
+
+        getLedgerDescCalledLatch = new CountDownLatch(1);
+        getLedgerDescWaitLatch = new CountDownLatch(1);
+
+        // spy `getLedgerForEntry`
+        doAnswer(invocationOnMock -> {
+            ByteBuf entry = invocationOnMock.getArgument(0);
+            long ledgerId = entry.getLong(entry.readerIndex());
+
+            LedgerDescriptor ld = (LedgerDescriptor) invocationOnMock.callRealMethod();
+
+            if (ledgerId % 2 == 1) {
+                getLedgerDescCalledLatch.countDown();
+                getLedgerDescWaitLatch.await();
+            }
+
+            return ld;
+        }).when(bookie).getLedgerForEntry(
+            any(ByteBuf.class),
+            any(byte[].class));
+    }
+
+    @After
+    public void teardown() throws Exception {
+        if (null != bookie) {
+            bookie.shutdown();
+        }
+    }
+
+    private static ByteBuf createByteBuf(long ledgerId, long entryId, int entrySize) {
+        byte[] data = new byte[entrySize];
+        ThreadLocalRandom.current().nextBytes(data);
+        ByteBuf buffer = Unpooled.wrappedBuffer(data);
+        buffer.writerIndex(0);
+        buffer.writeLong(ledgerId);
+        buffer.writeLong(entryId);
+        buffer.writeLong(entryId - 1); // lac
+        buffer.writerIndex(entrySize);
+        return buffer;
+    }
+
+    @Test
+    public void testCheckpoint() throws Exception {
```

Comment:
This test looks like you're trying to hit a very specific sequence of events and thread interleaving. It would be clearer to not use threads at all and directly call the interleaving you as looking for.

Reply:
You have to use threads, you can’t call directly. Because you need to bring in the race conditions to trigger this sequence, and you have to do that in a separate thread.

Reply:
I'm not so sure, but anyhow, as I said, non-blocking. This can be merged now.